File: python/features/steps/realtime_steps.py
"""Behave steps for realtime gossip and overlay."""

from __future__ import annotations

import json
import os
import socket
import tempfile
import threading
import time
from datetime import datetime, timezone
from pathlib import Path

from behave import given, then, when

from kanbus.gossip import DedupeSet, GossipEnvelope, ensure_mosquitto
from kanbus.models import IssueData, OverlayConfig
from kanbus.overlay import (
    OverlayIssueRecord,
    gc_overlay,
    overlay_issue_path,
    resolve_issue_with_overlay,
    write_overlay_issue,
)


def _parse_ts(ts: str) -> datetime:
    return datetime.fromisoformat(ts.replace("Z", "+00:00"))


def _issue(identifier: str, updated_at: datetime) -> IssueData:
    return IssueData(
        id=identifier,
        title="Realtime test",
        description="",
        type="task",
        status="open",
        priority=2,
        assignee=None,
        creator=None,
        parent=None,
        labels=[],
        dependencies=[],
        comments=[],
        created_at=updated_at,
        updated_at=updated_at,
        closed_at=None,
        custom={},
    )


@when("I read the realtime documentation")
def when_read_realtime_docs(context: object) -> None:
    root = Path(__file__).resolve().parents[3]
    doc_path = root / "docs" / "REALTIME.md"
    if not doc_path.exists():
        raise AssertionError("docs/REALTIME.md not found")
    context.realtime_doc = doc_path.read_text(encoding="utf-8")


@then("the realtime guide documents transport selection")
def then_doc_transport(context: object) -> None:
    assert "Transport selection" in context.realtime_doc


@then("the realtime guide documents broker discovery")
def then_doc_broker_discovery(context: object) -> None:
    assert "Discovery precedence" in context.realtime_doc


@then("the realtime guide documents autostart behavior")
def then_doc_autostart(context: object) -> None:
    assert "Autostart" in context.realtime_doc


@then("the realtime guide documents envelope schema")
def then_doc_envelope(context: object) -> None:
    assert "Envelope schema" in context.realtime_doc


@then("the realtime guide documents dedupe rules")
def then_doc_dedupe(context: object) -> None:
    assert "Dedupe" in context.realtime_doc


@then("the realtime guide documents overlay merge rules")
def then_doc_overlay_merge(context: object) -> None:
    assert "Overlay merge" in context.realtime_doc


@then("the realtime guide documents overlay GC and hooks")
def then_doc_overlay_gc(context: object) -> None:
    assert "overlay gc" in context.realtime_doc
    assert "install-hooks" in context.realtime_doc


@then("the realtime guide documents CLI commands")
def then_doc_cli(context: object) -> None:
    assert "gossip watch" in context.realtime_doc


@then("the realtime guide documents config blocks")
def then_doc_config(context: object) -> None:
    assert "realtime:" in context.realtime_doc
    assert "overlay:" in context.realtime_doc


@given('a gossip issue "{identifier}" updated at "{timestamp}"')
def given_gossip_issue(context: object, identifier: str, timestamp: str) -> None:
    updated_at = _parse_ts(timestamp)
    context.gossip_issue = _issue(identifier, updated_at)


@when("I build a gossip envelope for the issue")
def when_build_envelope(context: object) -> None:
    issue: IssueData = context.gossip_issue
    context.gossip_envelope = GossipEnvelope(
        id="n1",
        ts=issue.updated_at.isoformat().replace("+00:00", "Z"),
        project="kanbus",
        type="issue.mutated",
        issue_id=issue.identifier,
        event_id="evt-1",
        producer_id="producer-1",
        issue=issue,
    )


@then("the envelope includes the issue snapshot")
def then_envelope_includes_issue(context: object) -> None:
    envelope: GossipEnvelope = context.gossip_envelope
    payload = envelope.model_dump(by_alias=True, mode="json")
    assert payload.get("issue") is not None


@then("the envelope includes standard metadata fields")
def then_envelope_standard_metadata(context: object) -> None:
    envelope: GossipEnvelope = context.gossip_envelope
    payload = envelope.model_dump(by_alias=True, mode="json")
    for key in ("id", "ts", "project", "type", "issue_id", "event_id", "producer_id"):
        assert key in payload


@given('a gossip receiver with producer id "{producer_id}"')
def given_gossip_receiver(context: object, producer_id: str) -> None:
    context.gossip_producer_id = producer_id
    context.gossip_dedupe = DedupeSet(ttl_s=3600)


@given('it has already seen notification id "{notification_id}"')
def given_seen_notification(context: object, notification_id: str) -> None:
    context.gossip_dedupe.seen(notification_id)


@when(
    'it receives notification id "{notification_id}" from producer "{producer_id}"'
)
def when_receive_notification(
    context: object, notification_id: str, producer_id: str
) -> None:
    ignored = False
    if context.gossip_dedupe.seen(notification_id):
        ignored = True
    if producer_id == context.gossip_producer_id:
        ignored = True
    context.last_notification_ignored = ignored


@then("the notification is ignored")
def then_notification_ignored(context: object) -> None:
    assert context.last_notification_ignored is True


@given('a base issue "{identifier}" updated at "{timestamp}"')
def given_base_issue(context: object, identifier: str, timestamp: str) -> None:
    updated_at = _parse_ts(timestamp)
    context.overlay_base_issue = _issue(identifier, updated_at)
    if not getattr(context, "overlay_project_dir", None):
        temp_dir = tempfile.TemporaryDirectory()
        context.overlay_temp_dir = temp_dir
        project_dir = Path(temp_dir.name) / "project"
        (project_dir / "issues").mkdir(parents=True, exist_ok=True)
        context.overlay_project_dir = project_dir
    issue_path = context.overlay_project_dir / "issues" / f"{identifier}.json"
    issue_path.write_text(
        json.dumps(
            context.overlay_base_issue.model_dump(by_alias=True, mode="json"), indent=2
        ),
        encoding="utf-8",
    )


@given('an overlay issue "{identifier}" updated at "{timestamp}"')
def given_overlay_issue(context: object, identifier: str, timestamp: str) -> None:
    updated_at = _parse_ts(timestamp)
    overlay_issue = _issue(identifier, updated_at)
    context.overlay_issue_record = OverlayIssueRecord(
        issue=overlay_issue,
        overlay_ts=updated_at.isoformat().replace("+00:00", "Z"),
        overlay_event_id=None,
    )


@when("I resolve the overlay issue")
def when_resolve_overlay(context: object) -> None:
    project_dir: Path = context.overlay_project_dir
    context.overlay_resolved = resolve_issue_with_overlay(
        project_dir,
        context.overlay_base_issue,
        context.overlay_issue_record,
        None,
        OverlayConfig(enabled=True, ttl_s=86400),
    )


@then("the overlay version is returned")
def then_overlay_version(context: object) -> None:
    resolved = context.overlay_resolved
    assert resolved is not None
    assert resolved.updated_at == context.overlay_issue_record.issue.updated_at


@given('an overlay snapshot "{identifier}" updated at "{timestamp}"')
def given_overlay_snapshot(context: object, identifier: str, timestamp: str) -> None:
    updated_at = _parse_ts(timestamp)
    overlay_issue = _issue(identifier, updated_at)
    write_overlay_issue(
        context.overlay_project_dir,
        overlay_issue,
        updated_at.isoformat().replace("+00:00", "Z"),
        None,
    )


@when("I run overlay GC")
def when_run_overlay_gc(context: object) -> None:
    gc_overlay(context.overlay_project_dir, OverlayConfig(enabled=True, ttl_s=86400))


@then('the overlay snapshot "{identifier}" is removed')
def then_overlay_removed(context: object, identifier: str) -> None:
    assert not overlay_issue_path(context.overlay_project_dir, identifier).exists()


@given("a running UDS gossip broker")
def given_running_uds_broker(context: object) -> None:
    temp_dir = tempfile.TemporaryDirectory()
    context.uds_temp_dir = temp_dir
    socket_path = Path(temp_dir.name) / "bus.sock"
    context.uds_socket_path = socket_path

    def _run() -> None:
        from kanbus.gossip import run_uds_broker

        run_uds_broker(socket_path)

    thread = threading.Thread(target=_run, daemon=True)
    thread.start()
    context.uds_thread = thread
    time.sleep(0.1)


@when('a subscriber listens on "{topic}"')
def when_subscribe_uds(context: object, topic: str) -> None:
    sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
    sock.settimeout(2.0)
    sock.connect(str(context.uds_socket_path))
    payload = json.dumps({"op": "sub", "topic": topic}) + "\n"
    sock.sendall(payload.encode())
    context.uds_subscriber = sock
    context.uds_topic = topic


@when('a publisher sends a gossip envelope on "{topic}"')
def when_publish_uds(context: object, topic: str) -> None:
    issue = _issue("kanbus-uds", datetime.now(timezone.utc))
    envelope = GossipEnvelope(
        id="uds-msg-1",
        ts=issue.updated_at.isoformat().replace("+00:00", "Z"),
        project="kanbus",
        type="issue.mutated",
        issue_id=issue.identifier,
        event_id="evt-uds",
        producer_id="producer-uds",
        issue=issue,
    )
    context.uds_published_id = envelope.id
    payload = json.dumps(
        {"op": "pub", "topic": topic, "msg": envelope.model_dump(by_alias=True, mode="json")}
    ) + "\n"
    pub_sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
    pub_sock.settimeout(2.0)
    pub_sock.connect(str(context.uds_socket_path))
    pub_sock.sendall(payload.encode())
    pub_sock.close()


@then("the subscriber receives the envelope")
def then_subscriber_receives(context: object) -> None:
    sock: socket.socket = context.uds_subscriber
    buffer = b""
    while True:
        chunk = sock.recv(4096)
        if not chunk:
            break
        buffer += chunk
        if b"\n" in buffer:
            line, _rest = buffer.split(b"\n", 1)
            payload = json.loads(line.decode())
            msg = payload.get("msg", {})
            assert msg.get("id") == context.uds_published_id
            return
    raise AssertionError("subscriber did not receive envelope")


@given("mosquitto is available")
def given_mosquitto_available(context: object) -> None:
    if not _mosquitto_on_path():
        context.scenario.skip("mosquitto not installed")


@when("I autostart a mosquitto broker")
def when_autostart_mosquitto(context: object) -> None:
    from kanbus.gossip import _parse_broker_url

    endpoint = _parse_broker_url("mqtt://127.0.0.1:1883")
    startup = ensure_mosquitto(endpoint)
    context.mosquitto_startup = startup
    if startup is not None:
        # allow broker.json to flush
        time.sleep(0.1)


@then("broker metadata is written")
def then_broker_metadata_written(context: object) -> None:
    run_dir = Path.home() / ".kanbus" / "run"
    broker_path = run_dir / "broker.json"
    assert broker_path.exists()
    if context.mosquitto_startup is not None:
        context.mosquitto_startup.process.terminate()


def _mosquitto_on_path() -> bool:
    for entry in os.environ.get("PATH", "").split(":"):
        if not entry:
            continue
        candidate = Path(entry) / "mosquitto"
        if candidate.exists():
            return True
    return False

File: rust/features/steps/realtime_steps.rs
use std::fs;
use std::io::{BufRead, BufReader, Write};
use std::os::unix::net::UnixStream;
use std::path::{Path, PathBuf};
use std::thread;
use std::time::Duration;

use chrono::{DateTime, TimeZone, Utc};
use cucumber::{given, then, when};
use tempfile::TempDir;

use kanbus::gossip::{autostart_mosquitto, DedupeSet, GossipEnvelope};
use kanbus::models::{IssueData, OverlayConfig};
use kanbus::overlay::{
    gc_overlay, overlay_issue_path, resolve_issue_with_overlay, write_overlay_issue,
    OverlayIssueRecord,
};

use crate::step_definitions::initialization_steps::KanbusWorld;

fn parse_ts(value: &str) -> DateTime<Utc> {
    DateTime::parse_from_rfc3339(value)
        .map(|dt| dt.with_timezone(&Utc))
        .unwrap_or_else(|_| Utc.with_ymd_and_hms(2026, 3, 6, 0, 0, 0).unwrap())
}

fn issue(identifier: &str, updated_at: DateTime<Utc>) -> IssueData {
    IssueData {
        identifier: identifier.to_string(),
        title: "Realtime test".to_string(),
        description: String::new(),
        issue_type: "task".to_string(),
        status: "open".to_string(),
        priority: 2,
        assignee: None,
        creator: None,
        parent: None,
        labels: Vec::new(),
        dependencies: Vec::new(),
        comments: Vec::new(),
        created_at: updated_at,
        updated_at,
        closed_at: None,
        custom: std::collections::BTreeMap::new(),
    }
}

#[when("I read the realtime documentation")]
fn when_read_realtime_docs(world: &mut KanbusWorld) {
    let root = PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("..");
    let doc_path = root.join("docs").join("REALTIME.md");
    let contents = fs::read_to_string(&doc_path).expect("REALTIME doc");
    world.realtime_doc = Some(contents);
}

#[then("the realtime guide documents transport selection")]
fn then_doc_transport(world: &mut KanbusWorld) {
    assert!(world
        .realtime_doc
        .as_ref()
        .expect("doc")
        .contains("Transport selection"));
}

#[then("the realtime guide documents broker discovery")]
fn then_doc_broker(world: &mut KanbusWorld) {
    assert!(world
        .realtime_doc
        .as_ref()
        .expect("doc")
        .contains("Discovery precedence"));
}

#[then("the realtime guide documents autostart behavior")]
fn then_doc_autostart(world: &mut KanbusWorld) {
    assert!(world
        .realtime_doc
        .as_ref()
        .expect("doc")
        .contains("Autostart"));
}

#[then("the realtime guide documents envelope schema")]
fn then_doc_envelope(world: &mut KanbusWorld) {
    assert!(world
        .realtime_doc
        .as_ref()
        .expect("doc")
        .contains("Envelope schema"));
}

#[then("the realtime guide documents dedupe rules")]
fn then_doc_dedupe(world: &mut KanbusWorld) {
    assert!(world.realtime_doc.as_ref().expect("doc").contains("Dedupe"));
}

#[then("the realtime guide documents overlay merge rules")]
fn then_doc_overlay(world: &mut KanbusWorld) {
    assert!(world
        .realtime_doc
        .as_ref()
        .expect("doc")
        .contains("Overlay merge"));
}

#[then("the realtime guide documents overlay GC and hooks")]
fn then_doc_gc(world: &mut KanbusWorld) {
    let doc = world.realtime_doc.as_ref().expect("doc");
    assert!(doc.contains("overlay gc"));
    assert!(doc.contains("install-hooks"));
}

#[then("the realtime guide documents CLI commands")]
fn then_doc_cli(world: &mut KanbusWorld) {
    assert!(world
        .realtime_doc
        .as_ref()
        .expect("doc")
        .contains("gossip watch"));
}

#[then("the realtime guide documents config blocks")]
fn then_doc_config(world: &mut KanbusWorld) {
    let doc = world.realtime_doc.as_ref().expect("doc");
    assert!(doc.contains("realtime:"));
    assert!(doc.contains("overlay:"));
}

#[given(expr = "a gossip issue {string} updated at {string}")]
fn given_gossip_issue(world: &mut KanbusWorld, identifier: String, timestamp: String) {
    let updated_at = parse_ts(&timestamp);
    world.gossip_issue = Some(issue(&identifier, updated_at));
}

#[when("I build a gossip envelope for the issue")]
fn when_build_envelope(world: &mut KanbusWorld) {
    let issue = world.gossip_issue.clone().expect("issue");
    let envelope = GossipEnvelope {
        id: "n1".to_string(),
        ts: issue
            .updated_at
            .to_rfc3339_opts(chrono::SecondsFormat::Millis, true),
        project: "kanbus".to_string(),
        event_type: "issue.mutated".to_string(),
        issue_id: Some(issue.identifier.clone()),
        event_id: Some("evt-1".to_string()),
        producer_id: "producer-1".to_string(),
        origin_cluster_id: None,
        issue: Some(issue),
    };
    world.gossip_envelope = Some(envelope);
}

#[then("the envelope includes the issue snapshot")]
fn then_envelope_includes_issue(world: &mut KanbusWorld) {
    let envelope = world.gossip_envelope.as_ref().expect("envelope");
    let payload = serde_json::to_value(envelope).expect("serialize");
    assert!(payload.get("issue").is_some());
}

#[then("the envelope includes standard metadata fields")]
fn then_envelope_standard_metadata(world: &mut KanbusWorld) {
    let envelope = world.gossip_envelope.as_ref().expect("envelope");
    let payload = serde_json::to_value(envelope).expect("serialize");
    for key in [
        "id",
        "ts",
        "project",
        "type",
        "issue_id",
        "event_id",
        "producer_id",
    ] {
        assert!(payload.get(key).is_some(), "missing metadata field: {key}");
    }
}

#[given(expr = "a gossip receiver with producer id {string}")]
fn given_receiver(world: &mut KanbusWorld, producer_id: String) {
    world.gossip_producer_id = Some(producer_id);
    world.gossip_dedupe = Some(DedupeSet::new(Duration::from_secs(3600)));
}

#[given(expr = "it has already seen notification id {string}")]
fn given_seen(world: &mut KanbusWorld, notification_id: String) {
    if let Some(dedupe) = world.gossip_dedupe.as_mut() {
        dedupe.seen(&notification_id);
    }
}

#[when(expr = "it receives notification id {string} from producer {string}")]
fn when_receive(world: &mut KanbusWorld, notification_id: String, producer_id: String) {
    let mut ignored = false;
    if let Some(dedupe) = world.gossip_dedupe.as_mut() {
        if dedupe.seen(&notification_id) {
            ignored = true;
        }
    }
    if world.gossip_producer_id.as_deref() == Some(producer_id.as_str()) {
        ignored = true;
    }
    world.last_notification_ignored = Some(ignored);
}

#[then("the notification is ignored")]
fn then_notification_ignored(world: &mut KanbusWorld) {
    assert_eq!(world.last_notification_ignored, Some(true));
}

#[given(expr = "a base issue {string} updated at {string}")]
fn given_base_issue(world: &mut KanbusWorld, identifier: String, timestamp: String) {
    let updated_at = parse_ts(&timestamp);
    let base_issue = issue(&identifier, updated_at);
    if world.overlay_project_dir.is_none() {
        let temp_dir = TempDir::new().expect("tempdir");
        let project_dir = temp_dir.path().join("project");
        fs::create_dir_all(project_dir.join("issues")).expect("issues dir");
        world.overlay_temp_dir = Some(temp_dir);
        world.overlay_project_dir = Some(project_dir);
    }
    let project_dir = world.overlay_project_dir.as_ref().expect("project dir");
    let issue_path = project_dir
        .join("issues")
        .join(format!("{identifier}.json"));
    fs::write(
        &issue_path,
        serde_json::to_vec(&base_issue).expect("serialize"),
    )
    .expect("write issue");
    world.overlay_base_issue = Some(base_issue);
}

#[given(expr = "an overlay issue {string} updated at {string}")]
fn given_overlay_issue(world: &mut KanbusWorld, identifier: String, timestamp: String) {
    let updated_at = parse_ts(&timestamp);
    let overlay_issue = issue(&identifier, updated_at);
    world.overlay_issue_record = Some(OverlayIssueRecord {
        overlay_ts: updated_at.to_rfc3339_opts(chrono::SecondsFormat::Millis, true),
        overlay_event_id: None,
        overrides: None,
        issue: overlay_issue,
    });
}

#[when("I resolve the overlay issue")]
fn when_resolve_overlay(world: &mut KanbusWorld) {
    let project_dir = world.overlay_project_dir.as_ref().expect("project dir");
    let resolved = resolve_issue_with_overlay(
        project_dir,
        world.overlay_base_issue.clone(),
        world.overlay_issue_record.clone(),
        None,
        &OverlayConfig {
            enabled: true,
            ttl_s: 86_400,
        },
        None,
    )
    .expect("resolve");
    world.overlay_resolved = resolved;
}

#[then("the overlay version is returned")]
fn then_overlay_version(world: &mut KanbusWorld) {
    let resolved = world.overlay_resolved.as_ref().expect("resolved");
    let overlay = world.overlay_issue_record.as_ref().expect("overlay");
    assert_eq!(resolved.updated_at, overlay.issue.updated_at);
}

#[given(expr = "an overlay snapshot {string} updated at {string}")]
fn given_overlay_snapshot(world: &mut KanbusWorld, identifier: String, timestamp: String) {
    let updated_at = parse_ts(&timestamp);
    let overlay_issue = issue(&identifier, updated_at);
    let project_dir = world.overlay_project_dir.as_ref().expect("project dir");
    write_overlay_issue(
        project_dir,
        &overlay_issue,
        &updated_at.to_rfc3339_opts(chrono::SecondsFormat::Millis, true),
        None,
    )
    .expect("write overlay");
    world.overlay_snapshot_id = Some(identifier);
}

#[when("I run overlay GC")]
fn when_run_overlay_gc(world: &mut KanbusWorld) {
    let project_dir = world.overlay_project_dir.as_ref().expect("project dir");
    gc_overlay(
        project_dir,
        &OverlayConfig {
            enabled: true,
            ttl_s: 86_400,
        },
    )
    .expect("gc overlay");
}

#[then(expr = "the overlay snapshot {string} is removed")]
fn then_overlay_removed(world: &mut KanbusWorld, identifier: String) {
    let project_dir = world.overlay_project_dir.as_ref().expect("project dir");
    let path = overlay_issue_path(project_dir, &identifier);
    assert!(!path.exists());
}

#[given("a running UDS gossip broker")]
fn given_uds_broker(world: &mut KanbusWorld) {
    let temp_dir = TempDir::new().expect("tempdir");
    let socket_path = temp_dir.path().join("bus.sock");
    world.uds_temp_dir = Some(temp_dir);
    world.uds_socket_path = Some(socket_path.clone());
    thread::spawn(move || {
        let _ = kanbus::gossip::run_gossip_broker(Path::new("."), Some(socket_path));
    });
    thread::sleep(Duration::from_millis(100));
}

#[when(expr = "a subscriber listens on {string}")]
fn when_subscribe_uds(world: &mut KanbusWorld, topic: String) {
    let socket_path = world.uds_socket_path.clone().expect("socket");
    let mut stream = UnixStream::connect(socket_path).expect("connect");
    stream.set_read_timeout(Some(Duration::from_secs(2))).ok();
    let payload = serde_json::json!({"op": "sub", "topic": topic});
    let line = serde_json::to_string(&payload).expect("serialize") + "\n";
    stream.write_all(line.as_bytes()).expect("write");
    world.uds_subscriber = Some(stream);
}

#[when(expr = "a publisher sends a gossip envelope on {string}")]
fn when_publish_uds(world: &mut KanbusWorld, topic: String) {
    let socket_path = world.uds_socket_path.clone().expect("socket");
    let issue = issue("kanbus-uds", Utc::now());
    let envelope = GossipEnvelope {
        id: "uds-msg-1".to_string(),
        ts: issue
            .updated_at
            .to_rfc3339_opts(chrono::SecondsFormat::Millis, true),
        project: "kanbus".to_string(),
        event_type: "issue.mutated".to_string(),
        issue_id: Some(issue.identifier.clone()),
        event_id: Some("evt-uds".to_string()),
        producer_id: "producer-uds".to_string(),
        origin_cluster_id: None,
        issue: Some(issue),
    };
    world.uds_published_id = Some(envelope.id.clone());
    let payload = serde_json::json!({"op": "pub", "topic": topic, "msg": envelope});
    let line = serde_json::to_string(&payload).expect("serialize") + "\n";
    let mut stream = UnixStream::connect(socket_path).expect("connect");
    stream.write_all(line.as_bytes()).expect("write");
}

#[then("the subscriber receives the envelope")]
fn then_subscriber_receives(world: &mut KanbusWorld) {
    let mut stream = world.uds_subscriber.take().expect("subscriber");
    let mut reader = BufReader::new(&mut stream);
    let mut line = String::new();
    let _ = reader.read_line(&mut line).expect("read");
    let payload: serde_json::Value = serde_json::from_str(&line).expect("json");
    let msg = payload.get("msg").expect("msg");
    let received_id = msg.get("id").and_then(|value| value.as_str()).unwrap_or("");
    assert_eq!(Some(received_id.to_string()), world.uds_published_id);
}

#[given("mosquitto is available")]
fn given_mosquitto_available(_world: &mut KanbusWorld) {
    if !mosquitto_on_path() {
        eprintln!("mosquitto not installed; skipping scenario");
    }
}

#[when("I autostart a mosquitto broker")]
fn when_autostart_mosquitto(world: &mut KanbusWorld) {
    if !mosquitto_on_path() {
        return;
    }
    let startup = autostart_mosquitto("mqtt://127.0.0.1:1883").expect("autostart");
    world.mosquitto_startup = startup;
    thread::sleep(Duration::from_millis(100));
}

#[then("broker metadata is written")]
fn then_broker_metadata(world: &mut KanbusWorld) {
    if !mosquitto_on_path() {
        return;
    }
    let broker_path = std::env::var("HOME")
        .map(PathBuf::from)
        .unwrap_or_else(|_| PathBuf::from("."))
        .join(".kanbus")
        .join("run")
        .join("broker.json");
    assert!(broker_path.exists());
    if let Some(mut startup) = world.mosquitto_startup.take() {
        let _ = startup.process.kill();
    }
}

fn mosquitto_on_path() -> bool {
    if let Ok(path) = std::env::var("PATH") {
        for entry in std::env::split_paths(&path) {
            if entry.join("mosquitto").exists() {
                return true;
            }
        }
    }
    false
}
